### PR TITLE
Add end-to-end tests for Navigation block

### LINKED ISSUE #18451: Navigation Block: add automated tests
There should be some automated testing in place for the block as it transitions from an experimental block to a non-experimental one:

- Test fixtures for the 'full-content' test (https://github.com/WordPress/gutenberg/tree/master/packages/e2e-tests/fixtures/blocks)
- Block functional tests for various flows
- Unit tests for any new APIs added (if they're non-experimental).

### PR DESCRIPTION
**Pull request by talldan**
## Description
Closes #18451

Adds e2e tests for the Navigation Block.

## How has this been tested?
It is tests.

## Types of changes
Tests

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 1]
```diff
@@ -4,26 +4,51 @@
...
+
+		// Create an empty nav block.
 		const [ createEmptyButton ] = await page.$x( '//button[text()="Create empty"]' );
 		await createEmptyButton.click();
-		await page.waitForSelector( 'input[placeholder="Search or type url"]' );
-		await page.type( 'input[placeholder="Search or type url"]', 'http://example.com' );
 
+		// Add a link to the default Navigation Link block.
+		await page.type( 'input[placeholder="Search or type url"]', 'https://wordpress.org' );
 		await page.keyboard.press( 'Enter' );
```
Comment (talldan) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: This is currently not working in `master`.

File: packages/block-library/src/navigation/edit.js [hunk 2]
```diff
@@ -100,6 +100,14 @@ function Navigation( {
 	// indicated they want to automatically add top level Pages
 	// then show the Placeholder
 	if ( ! hasExistingNavItems ) {
+		if ( isRequestingPages ) {
```
Comment (talldan) [inline packages/block-library/src/navigation/edit.js]: TODO - move this into a separate PR.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 3]
```diff
@@ -4,26 +4,65 @@
...
+	getEditedPostContent,
 	insertBlock,
+	pressKeyWithModifier,
 } from '@wordpress/e2e-test-utils';
 
-describe( 'Adds Navigation links', () => {
+async function updateActiveNavigationLink( { url, label } ) {
+	if ( url ) {
+		await page.type( 'input[placeholder="Search or type url"]', url );
+		await page.keyboard.press( 'Enter' );
```
Comment (talldan) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: TODO - wait for suggestions to appear before pressing Enter.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 4]
```diff
@@ -4,26 +4,65 @@
...
 		await createEmptyButton.click();
-		await page.waitForSelector( 'input[placeholder="Search or type url"]' );
-		await page.type( 'input[placeholder="Search or type url"]', 'http://example.com' );
 
-		await page.keyboard.press( 'Enter' );
+		// Add a link to the default Navigation Link block.
+		await updateActiveNavigationLink( { url: 'https://wordpress.org', label: 'WP' } );
+
+		// Move the mouse to reveal the block movers. Without this the test seems to fail.
+		await page.mouse.move( 10, 10 );
```
Comment (talldan) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: TODO - why?

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 5]
```diff
@@ -4,26 +4,65 @@
...
+	it( 'allows a navigation menu to be created using existing pages', async () => {
+		// Add the navigation block.
 		await insertBlock( 'Navigation' );
+
+		// Create an empty nav block.
+		await page.waitForSelector( '.wp-block-navigation-placeholder' );
+		const [ createFromExistingButton ] = await page.$x( '//button[text()="Create from all top pages"]' );
+		await createFromExistingButton.click();
+
+		// Snapshot should contain the default 'Sample Page'.
```
Comment (talldan) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: TODO - investigate mocking the API request for pages.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 6]
```diff
@@ -54,9 +54,13 @@ async function updateActiveNavigationLink( { url, label } ) {
 		await page.waitForXPath( `//span[@class="block-editor-link-control__search-item-title"]/mark[text()="${ url }"]` );
 		await page.keyboard.press( 'Enter' );
 	}
+
 	if ( label ) {
 		await page.click( '.wp-block-navigation-link__content.is-selected' );
+		const textContent = await page.evaluate( () => document.activeElement.textContent );
+		expect( textContent ).toBe( url );
 		await pressKeyWithModifier( 'primary', 'a' );
```
Comment (talldan) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: No idea why, but the second time around the test fails to delete the text. 🤔 

This is only happening on travis, locally they pass everytime. 😢

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 7]
```diff
@@ -3,27 +3,143 @@
...
 		const [ createEmptyButton ] = await page.$x( '//button[text()="Create empty"]' );
 		await createEmptyButton.click();
-		await page.waitForSelector( 'input[placeholder="Search or type url"]' );
-		await page.type( 'input[placeholder="Search or type url"]', 'http://example.com' );
 
-		await page.keyboard.press( 'Enter' );
+		// Add a link to the default Navigation Link block.
+		await updateActiveNavigationLink( { url: 'https://wordpress.org', label: 'WP' } );
+
+		// Move the mouse to reveal the block movers. Without this the test seems to fail.
```
Comment (tellthemachines) [inline packages/e2e-tests/specs/editor/blocks/navigation.test.js]: That's weird, the element is definitely visible on the page and it's focusable too, wonder why `page.click` doesn't work 🤔

**Review by tellthemachines — APPROVED** [review]
Looks good! Left a couple minor comments/questions.

File: packages/e2e-test-utils/src/press-key-with-modifier.js [hunk 8]
```diff
@@ -75,7 +75,7 @@ async function emulateSelectAll() {
 				charCode: 0,
 				keyCode: isMac ? 93 : 17,
 				which: isMac ? 93 : 17,
-			} ),
+			} )
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/e2e-test-utils/src/press-key-with-modifier.js]: Only for arrays/objects, but this is a function parameter list. It wasn't a lint fix, but thought I'd tidy it while I was looking at it.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 9]
```diff
@@ -3,27 +3,143 @@
...
+		await page.keyboard.press( 'Enter' );
+	}
+
+	if ( label ) {
+		await page.click( '.wp-block-navigation-link__content.is-selected' );
+
+		// Ideally this would be `await pressKeyWithModifier( 'primary', 'a' )`
+		// to select all text like other tests do.
+		// Unfortunately, these tests don't seem to pass on Travis CI when
+		// using that approach, while using `Home` and `End` they do pass.
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: Yeah, really strange. The call to the function didn't have any effect so the snapshots were failing, where I expected the new text to replace the existing text, instead it was typed into the middle of the field:
"Contact Us" became "ContGet In Touchct Us", but was expected to be "Get In Touch".

Changing the label seems like an important part of the block, so I didn't want to exclude it from the test.

[the remark this answers is not in the document]
Comment (talldan) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: Again, worked fine locally, just travis that had the difficulties.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 10]
```diff
@@ -3,27 +3,143 @@
...
+			},
+		] );
+
+		// Add the navigation block.
+		await insertBlock( 'Navigation' );
+
+		// Create an empty nav block. The 'create' button is disabled until pages are loaded,
+		// so we must wait for it to become not-disabled.
+		await page.waitForXPath( '//button[text()="Create from all top pages"][not(@disabled)]' );
+		const [ createFromExistingButton ] = await page.$x( '//button[text()="Create from all top pages"][not(@disabled)]' );
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: The first function waits for the selector to be present. The next one selects the element so that it can be clicked.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js [hunk 11]
```diff
@@ -3,27 +3,143 @@
...
+		await createFromExistingButton.click();
+
+		// Snapshot should contain the mocked pages.
+		expect( await getEditedPostContent() ).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (talldan) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: The snapshots mean the test can check all the relevant attributes are being set correctly. The main shortcoming is that for dynamic blocks it doesn't actually check the markup in the post.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js (hunk 7 above)
Comment (talldan) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: Yep, there were a few curiosities with these tests 😄 

It looks like there's a chance the movers will be moved to the toolbar (#18685), so hopefully this won't be a problem for too much longer!

**Comment by talldan** [comment]
Thanks for reviewing @tellthemachines!

**Comment by ellatrix** [comment]
This is causing failing e2e tests in `master`.

File: packages/e2e-tests/specs/editor/blocks/navigation.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (tellthemachines) [reply packages/e2e-tests/specs/editor/blocks/navigation.test.js]: I wonder if the keyboard travis uses is different from what we expect. The bit that's failing is all about simulating keypresses.
